llvm/lib/Analysis/InstructionSimplify.cpp

```diff
@@ -4425,6 +4425,60 @@ Value *llvm::simplifyWithOpReplaced(Value *V, Value *Op, Value *RepOp,
                                   RecursionLimit);
 }
 
+/// Try simplifying the selection command when condition, trueValue, and
+/// falseValue are combinations of special bitwise operators.
+static Value *simplifySelectBitTestSpec(Value *CmpLHS, Value *CmpRHS,
+                                        Value *TrueVal, Value *FalseVal) {
+  Value *X, *Y;
+  // (X & Y) == 0 ? X | Y : X ^ Y   --> X ^ Y
+  // (X & Y) == 0 ? X ^ Y : X | Y   --> X | Y
+  // (X & Y) != 0 ? X | Y : X ^ Y   --> X | Y
+  // (X & Y) != 0 ? X ^ Y : X | Y   --> X ^ Y
+  if (match(CmpLHS, m_c_And(m_Value(X), m_Value(Y))) &&
+      match(CmpRHS, m_Zero())) {
+    if (match(TrueVal, m_c_Or(m_Specific(X), m_Specific(Y))) &&
+        match(FalseVal, m_c_Xor(m_Specific(X), m_Specific(Y)))) {
+      return FalseVal;
+    } else if (match(FalseVal, m_c_Or(m_Specific(X), m_Specific(Y))) &&
+               match(TrueVal, m_c_Xor(m_Specific(X), m_Specific(Y)))) {
+      return FalseVal;
+    }
+  }
+  // (X | Y) == 0 ? X & Y : X ^ Y   --> X ^ Y
+  // (X | Y) == 0 ? X ^ Y : X & Y   --> X & Y
+  // (X | Y) != 0 ? X & Y : X ^ Y   --> X & Y
+  // (X | Y) != 0 ? X ^ Y : X & Y   --> X ^ Y
+  if (match(CmpLHS, m_c_Or(m_Value(X), m_Value(Y))) &&
+      match(CmpRHS, m_Zero())) {
+    if (match(TrueVal, m_c_And(m_Specific(X), m_Specific(Y))) &&
+        match(FalseVal, m_c_Xor(m_Specific(X), m_Specific(Y)))) {
+      return FalseVal;
+    } else if (match(FalseVal, m_c_And(m_Specific(X), m_Specific(Y))) &&
+               match(TrueVal, m_c_Xor(m_Specific(X), m_Specific(Y)))) {
+      return FalseVal;
+    }
+  }
+  // (X ^ Y) == 0 ? X | Y : X & Y   --> X & Y
+  // (X ^ Y) == 0 ? X & Y : X | Y   --> X | Y
+  // (X ^ Y) != 0 ? X | Y : X & Y   --> X | Y
+  // (X ^ Y) != 0 ? X & Y : X | Y   --> X & Y
+  // (X ^ Y) == 0 could be transformed to X == Y
+  // (X ^ Y) != 0 could be transformed to X != Y
+  if (match(CmpLHS, m_CombineOr(m_c_Xor(m_Value(X), m_Value(Y)), m_Value(X)))) {
+    if (Y && match(CmpRHS, m_Zero()) || match(CmpRHS, m_Value(Y))) {
```

Comment:
Does `X ^ Y == Y` / `Y == 0` work? If not think you need to refactor to something like 


```
X = nullptr; Y = nullptr;
if (!(match(CmpRHS, m_Zero()) && match(CmpLHS, m_Xor(m_Value(X), m_Value(Y))) {
 X = CmpLHS;
 Y = CmpRHS;
}
```

Reply:
I want to handle "(X^Y)==0" and "X==Y". is my code not enough?

Reply:
You code will also hande `X == 0` and `X^Y == Y` which isn't what you want I think.

Reply:
how about this?
```
  if ((!match(CmpRHS, m_Zero()) && match(CmpLHS, m_Value(X)) &&
       match(CmpRHS, m_Value(Y))) ||
      (match(CmpRHS, m_Zero()) &&
       match(CmpLHS, m_Xor(m_Value(X), m_Value(Y))))) {
 ```

Reply:
sure that looks correct.

Reply:
fixed it! thanks for advise!